### PR TITLE
fix: fix update emqxConfig field causes blueGreenUpdate exception

### PR DESCRIPTION
**Pull request by Gala-R**

### CODE REVIEW

File: apis/apps/v1beta4/emqxbroker_webhook.go [hunk 1]
```diff
@@ -233,3 +240,19 @@ func validateVolumeClaimTemplates(new, old Emqx) error {
 	}
 	return nil
 }
+
+func validateEmqxConfig(new, old Emqx) error {
+	oldEmqxConfig := old.GetSpec().GetTemplate().Spec.EmqxContainer.EmqxConfig
+	newEmqxConfig := new.GetSpec().GetTemplate().Spec.EmqxContainer.EmqxConfig
+	if oldEmqxConfig["name"] != newEmqxConfig["name"] {
+		return errors.New("refuse to update the name field in emqxconfig")
```
[the remark this answers is not in the document]
Comment (Gala-R) [reply apis/apps/v1beta4/emqxbroker_webhook.go]: > 

Done

File: apis/apps/v1beta4/emqxbroker_webhook.go [hunk 2]
```diff
@@ -233,3 +240,19 @@ func validateVolumeClaimTemplates(new, old Emqx) error {
...
+func validateEmqxConfig(new, old Emqx) error {
+	oldEmqxConfig := old.GetSpec().GetTemplate().Spec.EmqxContainer.EmqxConfig
+	newEmqxConfig := new.GetSpec().GetTemplate().Spec.EmqxContainer.EmqxConfig
+	if oldEmqxConfig["name"] != newEmqxConfig["name"] {
+		return errors.New(`refuse to update the "name" field in ".spec.template.spec.emqxContainer.emqxConfig"`)
+	}
+	for k, oldValue := range oldEmqxConfig {
+		if strings.HasPrefix(k, "cluster") {
+			if newValue, ok := newEmqxConfig[k]; ok && newValue != oldValue {
+				return errors.New(`refuse to update the "^cluster.*" field in ".spec.template.spec.emqxContainer.emqxConfig"`)
```
Comment (Rory-Z) [inline apis/apps/v1beta4/emqxbroker_webhook.go:253]:
```suggestion
				return errors.New(`refuse to update the "^cluster.*$" field in ".spec.template.spec.emqxContainer.emqxConfig"`)
```

File: apis/apps/v1beta4/emqxbroker_webhook_test.go [hunk 3]
```diff
@@ -232,4 +240,23 @@ func TestBrokerValidateUpdate(t *testing.T) {
...
+		assert.Error(t, new.ValidateUpdate(old))
+
+		new.Spec.Template.Spec.EmqxContainer.EmqxConfig["name"] = "emqx"
+		new.Spec.Template.Spec.EmqxContainer.EmqxConfig["cluster.dns.app"] = "emqx-test"
+		assert.Error(t, new.ValidateUpdate(old))
+
+		new.Spec.Template.Spec.EmqxContainer.EmqxConfig["name"] = "emqx"
+		new.Spec.Template.Spec.EmqxContainer.EmqxConfig["cluster.dns.app"] = "emqx"
+		new.Spec.Template.Spec.EmqxContainer.EmqxConfig["listener.tcp.internal"] = "0.0.0.0:1884"
+		assert.Nil(t, new.ValidateUpdate(old))
```
Comment (Rory-Z) [inline apis/apps/v1beta4/emqxbroker_webhook_test.go:260]: Check if the new configuration can be updated successfully if it does not contain name and cluster, just add new filed or just change `listener.tcp.internal`

File: apis/apps/v1beta4/emqxbroker_webhook.go (hunk 2 above)
Comment (Gala-R) [reply apis/apps/v1beta4/emqxbroker_webhook.go]: > 

What is the significance of adding the $ sign, our string does not have an end constraint

File: apis/apps/v1beta4/emqxbroker_webhook_test.go (hunk 3 above)
Comment (Gala-R) [reply apis/apps/v1beta4/emqxbroker_webhook_test.go:260]: > Check if the new configuration can be updated successfully if it does not contain name and cluster, just add new filed or just change `listener.tcp.internal`

A test case for the above has been added

**Review by Rory-Z — APPROVED** [review]

**Pull request merged**